src/renderer/base/renderer.cpp

```diff
@@ -64,44 +64,90 @@ Renderer::~Renderer()
-            if (SUCCEEDED(hr))
+        if (--tries == 0)
+        {
+            // Stop trying.
+            _pThread->DisablePainting();
+            if (_pfnRendererEnteredErrorState)
             {
-                break;
+                _pfnRendererEnteredErrorState();
             }
+            // If there's no callback, we still don't want to FAIL_FAST: the renderer going black
+            // isn't near as bad as the entire application aborting. We're a component. We shouldn't
+            // abort applications that host us.
+            return S_FALSE;
+        }
+
+        // Add a bit of backoff.
+        // Sleep 150ms, 300ms, 450ms before failing out and disabling the renderer.
+        Sleep(renderBackoffBaseTimeMilliseconds * (maxRetriesForRenderEngine - tries));
+    }
 
-            LOG_HR_IF(hr, hr != E_PENDING);
+    return S_OK;
+}
 
-            if (--tries == 0)
+[[nodiscard]] HRESULT Renderer::_PaintFrame() noexcept
+{
+    {
+        _pData->LockConsole();
+        auto unlock = wil::scope_exit([&]() {
+            _pData->UnlockConsole();
+        });
+
+        // Last chance check if anything scrolled without an explicit invalidate notification since the last frame.
+        _CheckViewportAndScroll();
+
+        if (_currentCursorOptions)
+        {
+            const auto coord = _currentCursorOptions->coordCursor;
+            const auto& buffer = _pData->GetTextBuffer();
+            const auto lineRendition = buffer.GetLineRendition(coord.y);
+            const auto cursorWidth = _pData->IsCursorDoubleWidth() ? 2 : 1;
+
+            til::rect cursorRect{ coord.x, coord.y, coord.x + cursorWidth, coord.y + 1 };
+            cursorRect = BufferToScreenLine(cursorRect, lineRendition);
+
+            if (buffer.GetSize().TrimToViewport(&cursorRect))
             {
-                // Stop trying.
-                _pThread->DisablePainting();
-                if (_pfnRendererEnteredErrorState)
+                FOREACH_ENGINE(pEngine)
                 {
-                    _pfnRendererEnteredErrorState();
+                    LOG_IF_FAILED(pEngine->InvalidateCursor(&cursorRect));
                 }
-                // If there's no callback, we still don't want to FAIL_FAST: the renderer going black
-                // isn't near as bad as the entire application aborting. We're a component. We shouldn't
-                // abort applications that host us.
-                return S_FALSE;
             }
+        }
 
-            // Add a bit of backoff.
-            // Sleep 150ms, 300ms, 450ms before failing out and disabling the renderer.
-            Sleep(renderBackoffBaseTimeMilliseconds * (maxRetriesForRenderEngine - tries));
+        _currentCursorOptions = _GetCursorInfo();
+
+        FOREACH_ENGINE(pEngine)
+        {
+            RETURN_IF_FAILED(_PaintFrameForEngine(pEngine));
         }
     }
 
+    FOREACH_ENGINE(pEngine)
+    {
+        RETURN_IF_FAILED(pEngine->Present());

```

Comment:
one big functional change here is that we now Present under lock. We did not do that before - theoretically it allowed us to yeet bits at the GPU (or whatever) while the console continued working. It was built on the idea that the slow operation would be finalization.

Is this no longer required? Is there a risk to making this locking change?

Reply:
The `_pData->LockConsole();` call (and unlock) is inside an extra scope. So this will still run without the lock being held.